Add functional Autoassess diagnostics: land surface metrics: permafrost, soil moisture, surface radiation

valeriupredoi opened the pull request:
<!--
    Thank you for contributing to our project!

    Please do not delete this text completely, but read the text below and keep
    items that seem relevant. If in doubt, just keep everything and add your
    own text at the top, a reviewer will update the checklist for you.

    While the checklist is intended to be filled in by the technical and scientific
    reviewers, it is the responsibility of the author of the pull request to make
    sure all items on it are properly implemented.
-->

## Description

<!--
    Please describe your changes here, especially focusing on why this pull request makes
    ESMValTool better and what problem it solves.

    Before you start, please read our contribution guidelines: https://docs.esmvaltool.org/en/latest/community/

    Please fill in the GitHub issue that is closed by this pull request, e.g. Closes #1903
-->
- Addresses part of #1931
- Link to documentation: https://esmvaltool--2170.org.readthedocs.build/en/2170/

* * *

## Before you get started

<!--
    Please discuss your idea with the development team before getting started,
    to avoid disappointment or unnecessary work later. The way to do this is
    to open a new issue on GitHub.
-->

- [x] [☝ Create an issue](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#contributing-code-and-documentation) to discuss what you are going to do - partially addresses #1931 

## Checklist

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

<!-- The next two lines turn the 🛠 and 🧪 below into hyperlinks -->
[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🛠][1] This pull request has a [descriptive title](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#pull-request-title)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#code-quality)
- [x] [🛠][1] [Documentation](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#documentation) is available
- [x] [🛠][1] [Tests](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#tests) run successfully
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#list-of-authors) is up to date
- [x] [🛠][1] Any changed dependencies have been [added or removed correctly](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#dependencies)
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#pull-request-checks) were successful

### [New or updated recipe/diagnostic](https://docs.esmvaltool.org/en/latest/community/diagnostic.html)

- [x] [🧪][2] [Recipe runs successfully](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#testing-recipes)
- [x] [🧪][2] [Recipe is well documented](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#recipe-and-diagnostic-documentation)
- [x] [🧪][2] [Figure(s) and data](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#diagnostic-output) look as expected from literature
- [x] [🛠][1] [Provenance information](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#recording-provenance) has been added

### [New or updated data reformatting script](https://docs.esmvaltool.org/en/latest/develop/dataset.html)

- [x] [🛠][1] [Documentation](https://docs.esmvaltool.org/en/latest/community/dataset.html#dataset-documentation) is available
- [x] [🛠][1] The dataset has been [added to the CMOR check recipe](https://docs.esmvaltool.org/en/latest/community/dataset.html#testing)
- [x] [🧪][2] Numbers and units of the data look [physically meaningful](https://docs.esmvaltool.org/en/latest/community/dataset.html#scientific-sanity-check)

***

To help with the number of pull requests:

-  🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValTool/pulls) in this repository

<!--
If you need help with any of the items on the checklists above, please do not hesitate to ask by commenting in the issue or pull request.
-->

## Autoassess `landsurface` metrics/diagnostics

Battle plan to include the Autassess `landsurface` recipes in ESMValTool v2.3:

- [x] resurrect them from the now extremely old `all_autoassess_recipes` branch into a new branch (this one) and create Draft PR
- [x] test run them on JASMIN with the latest dev ESMValCore and obtain results, stock datasets from recipe
- [x] rerun them with UKESM1 data for comparisons with PP data -> @alistairsellar needs to tell me the data specs
- [x] store results somewhere easily available for checking
- [x] run the equivalent Autoassess ones on UKESM pp data
- [x] compare and do scientific review
- [x] complete documentation
- [x] fix CMIP5 (mrlsl) - possibly write a CMIP5-specific recipe or recipe section
- [x] remove MOHC test recipes
- [x] add missing plots in gallery (from `permafrost` only, no other plots outputted)
- [ ] mergy-merge and grab a brewski :beer:

valeriupredoi commented:
OK - smashing news: I ran all four `landsurface` recipes with the latest esmval- code from current development branches (both Core and current Tool branch, this one). Specs are:

*esmval code*
```
(release220) [valeriu@sci1 esmvaltool_var_test]$ conda list esmval
# packages in environment at /home/users/valeriu/anaconda3/envs/release220:
#
# Name                    Version                   Build  Channel
esmvalcore                2.2.0                     dev_0    <develop>
esmvaltool                2.2.0                     dev_0    <develop>
esmvaltool-sample-data    0.0.3                    pypi_0    pypi
```

*Python, conda, iris and matplotlib*
```
(release220) [valeriu@sci1 esmvaltool_var_test]$ conda list python
# packages in environment at /home/users/valeriu/anaconda3/envs/release220:
#
# Name                    Version                   Build  Channel
antlr-python-runtime      4.7.2           py39hde42818_1002    conda-forge
msgpack-python            1.0.2            py39h1a9c180_1    conda-forge
python                    3.9.2           hffdb5ce_0_cpython    conda-forge
python-dateutil           2.8.1                      py_0    conda-forge
python-stratify           0.1.1           py39hce5d2b2_1003    conda-forge
python-xxhash             2.0.0            py39h3811e60_1    conda-forge
python_abi                3.9                      1_cp39    conda-forge

(release220) [valeriu@sci1 esmvaltool_var_test]$ conda -V
conda 4.10.1

(release220) [valeriu@sci1 esmvaltool_var_test]$ conda list iris
# packages in environment at /home/users/valeriu/anaconda3/envs/release220:
#
# Name                    Version                   Build  Channel
iris                      3.0.1            py39hf3d152e_1    conda-forge
(release220) [valeriu@sci1 esmvaltool_var_test]$ conda list matplotlib
# packages in environment at /home/users/valeriu/anaconda3/envs/release220:
#
# Name                    Version                   Build  Channel
matplotlib                3.3.4            py39hf3d152e_0    conda-forge
matplotlib-base           3.3.4            py39h2fa2bec_0    conda-forge
```

valeriupredoi commented:
@alistairsellar could you please give me the UKESM1 data specifications that I should be rerunning these recipes with? If you look at the `autoassess_landsurface_*` recipes in the current [recipes](https://github.com/ESMValGroup/ESMValTool/tree/autoassess_landsurface_recipes/esmvaltool/recipes) dir, none of the recipes use UKESM1, so if you could tell me what the exact datasets per recipe should be so we can do the comparisons with the PP data, that's be grand! :beer:

alistairsellar commented:
Hi V, please use these runs to allow direct comparison:
- HadGEM3-GC31-LL r1i1p1f3
- UKESM1-0-LL r5i1p1f3

The period you have already is fine: start_year: 1992,  end_year: 2002
but can I just check that this means 1992-01-01 to 2002-12-31

valeriupredoi commented:
excellent, cheers @alistairsellar :beer: are the other data parameters and variables good?

alistairsellar commented:
Yes, they look correct to me.

valeriupredoi commented:
excellent @alistairsellar - I'll run those tomorrow after 10am (JASMIN is having FS updates 830am-10am tomorrow, I can have coffee :coffee: ) :beer:

alistairsellar commented on esmvaltool/recipes/recipe_autoassess_landsurface_permafrost.yml:
```diff
@@ -0,0 +1,62 @@
+        mip: Lmon
+      sftlf:
+        mip: fx
+      mrsofc:
```
Actually, this should be mrsos (also in Lmon). The code wants the actual soil moisture content (=mrsos) whereas mrsofc is the soil moisture capacity.

valeriupredoi replied:
the CMIP6 CMOR table entry for `mrsos` says:
```
        "mrsos": {
            "frequency": "mon",
            "modeling_realm": "land",
            "standard_name": "mass_content_of_water_in_soil_layer",
            "units": "kg m-2",
            "cell_methods": "area: mean where land time: mean",
            "cell_measures": "area: areacella",
            "long_name": "Moisture in Upper Portion of Soil Column",
            "comment": "The mass of water in all phases in the upper 10cm of the  soil layer.",
            "dimensions": "longitude latitude time sdepth1",
            "out_name": "mrsos",
            "type": "real",
            "positive": "",
            "valid_min": "",
            "valid_max": "",
            "ok_min_mean_abs": "",
            "ok_max_mean_abs": ""
        },
        
 ```
are we sure that's what we need? I am changing the recipe now but that means changing the dig too

valeriupredoi replied:
another massive hurdle is that HadGEM is completely missing `mrsos` data from *all* experiments on JASMIN

alistairsellar replied:
Yes, definitely. The code comments are consistent with this, and the original stash code m01s08i223 maps to mrsos in our CMIP6 submission.

alistairsellar replied:
What's the "dig"?

valeriupredoi replied:
what dig?

alistairsellar replied:
You said
> that means changing the dig too

I didn't know what that meant.

valeriupredoi replied:
sorry - typo, should be diag -> diagnostic :laughing:

alistairsellar replied:
Haha - I should have worked that out.

alistairsellar commented on esmvaltool/recipes/recipe_autoassess_landsurface_surfrad.yml:
```diff
@@ -0,0 +1,75 @@
+  aa_landsurf_surfrad:
+    description: Autoassess test diag for Land-Surface Surfrad.
+    variables:
+      rsns: # Surf SW net all sky
```
I think this should be rss instead of rsns. In fact I can't find rsns in the CMIP6 variable names - is this a name that changed from CMIP5 to CMIP6?

valeriupredoi replied:
`rsns` is a derived variable with its own [custom table(https://github.com/ESMValGroup/ESMValCore/blob/master/esmvalcore/cmor/tables/custom/CMOR_rsns.dat) and [derivation script](https://github.com/ESMValGroup/ESMValCore/blob/master/esmvalcore/preprocessor/_derive/rsns.py) - `rss` is this guy:
```
        "rss": {
            "frequency": "mon",
            "modeling_realm": "atmos",
            "standard_name": "surface_net_downward_shortwave_flux",
            "units": "W m-2",
            "cell_methods": "area: time: mean",
            "cell_measures": "area: areacella",
            "long_name": "Net Shortwave Surface Radiation",
            "comment": "Net downward shortwave radiation at the surface",
            "dimensions": "longitude latitude time",
            "out_name": "rss",
            "type": "real",
            "positive": "down",
            "valid_min": "",
            "valid_max": "",
            "ok_min_mean_abs": "",
            "ok_max_mean_abs": ""
        },
        
```
and is Emon instead of Amon

alistairsellar replied:
Ah, OK - the derived variable is fine. It's exactly the same as rss. I'm guessing that rss didn't exist at CMIP5 and so the derived variable was needed.

valeriupredoi replied:
cool! cheers Ali :beer:

alistairsellar commented on esmvaltool/recipes/recipe_autoassess_landsurface_surfrad.yml:
```diff
@@ -0,0 +1,75 @@
+          - {dataset: CERES-EBAF,  project: obs4mips,  level: L3B,  version: Ed2-7,  start_year: 2001,  end_year: 2012, tier: 1}
+      sftlf:
+        mip: fx
+      rlns: # Surf LW net all sky
```
Similarly, at least for CMIP6 this should be rls instead of rlns.

valeriupredoi replied:
similarily too, this is [derived](https://github.com/ESMValGroup/ESMValCore/blob/master/esmvalcore/preprocessor/_derive/rlns.py) and it's got its own custom [table](https://github.com/ESMValGroup/ESMValCore/blob/master/esmvalcore/cmor/tables/custom/CMOR_rlns.dat) - I think we discussed these back in the day. Do you think these two need to change?

alistairsellar replied:
As above, the derived variable rnls is fine and is the same thing as the CMIP6 variable rls.

valeriupredoi replied:
OK excellent! :beer:

alistairsellar commented:
Hi again. I've now been through the original code in more detail, and picked up 3 changes to variable names needed in the recipes (see comments above).

I also noticed your comment about the missing soil moisture variables. I'm consulting with our local expert on this.

valeriupredoi commented:
Data issues: CMIP6 data is patchy on JASMIN: so far I have noticed `mrsos` data missing for HadGEM3-GC31-LL for experiments, and other bits and bobs missing for the setup you asked for @alistairsellar - for permafrost I could get data (apart for `mrsos` if I change to historical or piControl, would either of those do?

alistairsellar commented:
Any experiment with HadGEM3 or UKESM1 will be fine - I can make the equivalent AutoAssess in a couple of hours. You could even compare two UKESM exps, or two HadGEM3 exps if that's more complete.

valeriupredoi commented:
@alistairsellar can we not do it for the exact same model and experiment - comparing the individual metrics should probably do the trick? At least for `permafrost` `mrasos` is giving me terrible headaches for HadGEM missing everywhere. If I do two different experiments for UKESM1 that's no good either because the fx files (sftlf) lives only in piControl (this is a massive gripe me and @zklaus and @ledm have been having, replication of fx files across experiments is nonexistent in CMIP6 UKESM1 data)

alistairsellar commented:
Sure, that's fine. Whatever you can make work for available CMIP6 data, I can replicate here.

alistairsellar commented:
> this is a massive gripe...
Is fx thing this a Met Office-specific issue? If so, do you want me to pick it up with our data publishing experts?

valeriupredoi commented:
OK @alistairsellar I did manage to run the permafrost diag (not dig :rofl: ) with full data complement but for a very very restricted data call, see the [recipe](https://github.com/ESMValGroup/ESMValTool/blob/autoassess_landsurface_recipes/esmvaltool/recipes/recipe_autoassess_landsurface_permafrost_MOHC.yml) I used, if you could run the same I have the results at hand :+1:

valeriupredoi commented:
> > this is a massive gripe...
> 
> Is fx thing this a Met Office-specific issue? If so, do you want me to pick it up with our data publishing experts?

we spoke to @matthew-mizielinski and the take was (at the time we spoke, about a year ago) that these files, since they are identical across exps, should stay in one exp only, so not a serious problem but just a local annoyance :grin:

alistairsellar commented:
> if you could run the same I have the results at hand

Great! Setting it off now...

alistairsellar commented:
> since they are identical across exps, should stay in one exp only

Do other modelling centres take the same view? Or do they replicate fx files across all exps?

valeriupredoi commented:
**Permafrost RESULTS**
*Recipe*
[recipe](https://github.com/ESMValGroup/ESMValTool/blob/autoassess_landsurface_recipes/esmvaltool/recipes/recipe_autoassess_landsurface_permafrost_MOHC.yml)

*Metrics: UKESM1*
```
permafrost area,1354.5278365343268
fraction area permafrost over zerodeg,29.150328299094394
offset 1m minus surface,1.275604248046875
offset surface minus air,8.726593017578125
attenuation 1m over surface,0.49492978649195407
attenuation surface over air,0.4920951365355539
```

*Plots*

![pf_extent_asia_UKESM1-0-LL](https://user-images.githubusercontent.com/28983971/118290428-56f39a00-b4ce-11eb-9d37-bf4b074e8ad0.png)
![pf_extent_north_america_UKESM1-0-LL](https://user-images.githubusercontent.com/28983971/118290451-5d821180-b4ce-11eb-9339-756398eb4645.png)

alistairsellar commented:
Great stuff! That's a damn good start. I'm still waiting for my piControl AutoAssess to run, but doing an initial comparison to hist results: http://gws-access.jasmin.ac.uk/public/mo_gc3/dev2/autoassess/u-az513_vs_u-bg466/land_surface.html (should be visible to you if you're going in through Reading IP address - let me know if not).

For this rough comparison, the plots look very very good. However, the first two metrics too large by about 2 orders of magnitude. Could there be a mixup between fraction and % in what the code is expecting? I'll take a look...

valeriupredoi commented:
not sure yet, but there are two problems I can already see in my run:

- first off, the `mrsos` file we have off ESGF has depth coord at one single pint, and that is well below 2m:
```
/badc/cmip6/data/CMIP6/CMIP/MOHC/UKESM1-0-LL/piControl/r1i1p1f2/Lmon/mrsos/gn/latest/mrsos_Lmon_UKESM1-0-LL_piControl_r1i1p1f2_gn_196001-204912.nc
Depth coordinate:
DimCoord(array([0.05]), bounds=array([[0. , 0.1]]), standard_name='depth', units=Unit('m'), long_name='depth', var_name='depth', attributes={'positive': 'down'})
```
- the second issue is that `mrsos` is time varying, and the Autoassess code needs it to be 2d lat-lon only, so to do that I had to collapse time on MEAN, which is a hacky hack

valeriupredoi commented:
> you if you're going in through Reading IP address

at home today so I have a Reading West IP :grin: - so no, I can't sees it. Anyways, off the grid for a bit, need lunch :hamburger:

alistairsellar commented:
> it's that depth that's killing us

Sorry, I hadn't spotted that mrsos is just the shallowest soil layer. After more hunting I see that mrsol is the 3D equivalent and has 4 layers for the UK models, with the deepest at 2m.

alistairsellar commented:
And a time mean is correct - the original code uses a mean over the whole period.

alistairsellar commented:
> Reading West IP

We definitely don't put *those* on the allow-list! I'll email you the password...

valeriupredoi commented:
> > it's that depth that's killing us
> 
> Sorry, I hadn't spotted that mrsos is just the shallowest soil layer. After more hunting I see that mrsol is the 3D equivalent and has 4 layers for the UK models, with the deepest at 2m.

very cool! lemme try `mrsol` then

alistairsellar commented:
Ooh, also mrsol has actually been published for HadGEM3-GC31-LL piControl

valeriupredoi commented:
> Ooh, also mrsol has actually been published for HadGEM3-GC31-LL piControl

I can rerun if I find the data, just shipped the run for UKESM1 only for now

alistairsellar commented:
That's all we need to verify the metrics anyway.

valeriupredoi commented:
OK I just reran with `mrsol` and extracting the 2m layer and...the metrics are identical down to the last digit :worried:

alistairsellar commented:
Hrrm, that's a shame. 

I have an idea - looking into this idea now...

valeriupredoi commented:
just saw the plots you linked, they look almost identical to mines, but look at the scale, yours are 0-1, mines are 0-100, that may explain the two orders of magnitude (BTW I don't see the metrics values anywhere?)

alistairsellar commented:
In the meantime, here are the AutoAssess permafrost metrics for UKESM1 piControl 1997/12/01- 2002/12/01:
```
attenuation 1m over surface,0.49573169134648043
attenuation surface over air,0.4925018141311794
fraction area permafrost over zerodeg,0.28940865677944033
offset 1m minus surface,0.7005462646484375
offset surface minus air,5.88677978515625
permafrost area,13.561272172616677
```

valeriupredoi commented:
OK clearly there is an extra factor of 100 in some of my results, divide those by 100 and we get pretty dam close numbers

alistairsellar commented:
> I have an idea - looking into this idea now...

The idea came to naught. I might need to run with some debugging prints. Is it possible for me to run your recipe + branch on Jasmin?

valeriupredoi commented:
yes, I'll set up an executable script for you, but not now, I'm off for the day, will do it Monday first thing. But look at the metrics - they are almost identical (well, if you divide some of mines by 100)

alistairsellar commented:
Yes, it looks really really promising! 

Have a great weekend!

alistairsellar commented:
FWIW, the piControl (aw310) permafrost plots should appear here soon:
http://gws-access.jasmin.ac.uk/public/mo_gc3/dev2/autoassess/u-az513_vs_u-aw310/land_surface.html

matthew-mizielinski commented:
> > > this is a massive gripe...
> > 
> > 
> > Is fx thing this a Met Office-specific issue? If so, do you want me to pick it up with our data publishing experts?
> 
> we spoke to @matthew-mizielinski and the take was (at the time we spoke, about a year ago) that these files, since they are identical across exps, should stay in one exp only, so not a serious problem but just a local annoyance 😁

Hi @valeriupredoi, @alistairsellar 

I'll have a think about whether we can reconsider this, but a quick back of the envelope calculation gives ~3,000 simulations* x 9 fixed field per simulation = ~27,000 datasets which will all be duplicates.

The concept of a symbolic link within ESGF has been discussed, which would allow us to avoid duplication like this, but I don't expect progress in the near term.

*DCPP & PAMIP account for ~85% of this

alistairsellar commented:
Thanks @matthew-mizielinski. What have other modelling centres done in this regard? Or is it the usual hodge-podge of different approaches?

matthew-mizielinski commented:
> Thanks @matthew-mizielinski. What have other modelling centres done in this regard? Or is it the usual hodge-podge of different approaches?

If you do a search for `fx/orog` on esgf you can get a feels for the number. I think about a third have published it for everything, another third have a minimal set (including MOHC), and a third have a random selection.

valeriupredoi commented:
cheers @matthew-mizielinski - don't forget the third that have not published it at all :grin:

valeriupredoi commented:
> I'll have a think about whether we can reconsider this, but a quick back of the envelope calculation gives ~3,000 simulations* x 9 fixed field per simulation = ~27,000 datasets which will all be duplicates.

Indeed, it's a bit silly to replicate duplicates, but maybe include them at top ensemble (r1i1p1fX?) for each experiment? The CMIP5 r0i0p0 method was better IMHO :+1:

matthew-mizielinski commented:
> cheers @matthew-mizielinski - don't forget the third that have not published it at all 😁

Good point!

> > I'll have a think about whether we can reconsider this, but a quick back of the envelope calculation gives ~3,000 simulations* x 9 fixed field per simulation = ~27,000 datasets which will all be duplicates.
> 
> Indeed, it's a bit silly to replicate duplicates, but maybe include them at top ensemble (r1i1p1fX?) for each experiment? The CMIP5 r0i0p0 method was better IMHO 👍

Not a bad idea actually, this brings the number of required datasets down to below 2,500.

alistairsellar commented:
> maybe include them at top ensemble (r1i1p1fX?) for each experiment

Would that enable ESMValTool to pick them up for a different ensemble member of the same experiment? If not, I guess it would at least allow use of r1.

valeriupredoi commented:
cheers again @matthew-mizielinski - feel free to untag yourself from this PR, Alistair and myself will be exchanging quite a lot of communication here and I don't want you to get pinged everytime we post stuffs about Autoassess :grin: 

@alistairsellar I found out where we have that extra factor of 100 - using an `sftlf` mask that is in percents x 100, difiding the result of `pf_periods` by 100 to account for absulte land fraction (and not percent) gives me these metrics:
```
permafrost area,13.54527836534327
fraction area permafrost over zerodeg,0.29150328299094397
offset 1m minus surface,1.275604248046875
offset surface minus air,8.726593017578125
attenuation 1m over surface,0.49492978649195407
attenuation surface over air,0.4920951365355539
```
which are very very close to your UKESM1 '97-'02 period (great period that one BTW, some of the best years in me life haha)
```
attenuation 1m over surface,0.49573169134648043
attenuation surface over air,0.4925018141311794
fraction area permafrost over zerodeg,0.28940865677944033
offset 1m minus surface,0.7005462646484375
offset surface minus air,5.88677978515625
permafrost area,13.561272172616677
```

valeriupredoi commented:
the plots look sane and almost identical to yours as well:

![pf_extent_asia](https://user-images.githubusercontent.com/28983971/118476209-2184c180-b705-11eb-8558-5dc625b15abc.png)
![pf_extent_NA](https://user-images.githubusercontent.com/28983971/118476237-28abcf80-b705-11eb-95fd-2338a00dd2d2.png)

alistairsellar commented:
Brilliant! Well spotted.

So I think we're scientifically verified for:
 - Permafrost plots
 - attenuation 1m over surface
 - attenuation surface over air
 - fraction area permafrost over zerodeg,0.28940865677944033
 - permafrost area,13.561272172616677

But these two are still different:
 - offset 1m minus surface, 0.70 vs 1.3
 - offset surface minus air, 5.9 vs 8.7

I'll look at the code for these after lunch...

valeriupredoi commented:
awesomepossum! Naively those two metrics differ since the analysis periods differ - my analysis is 1992-2002 and yours is 1997-2002 I can run with 1997 right now :+1:

valeriupredoi commented:
> awesomepossum! Naively those two metrics differ since the analysis periods differ - my analysis is 1992-2002 and yours is 1997-2002 I can run with 1997 right now +1

Nevermind, I just reran with 97-02 and I got pretty much the same metrics, yeah maybe have a look at the code, beats me why they'd be different, bon appetit :+1:

valeriupredoi commented:
@alistairsellar I've realized I was not applying the median in masked mode (hence shifting counts with masked elements) and I have also changed the interpolator for the case when we extract layer 0 to Nearest since that makes more sense given that the data doesn't actually have a layer=0, so with these changes I got:
```
offset 1m minus surface,0.86416626
offset surface minus air,7.6981506
```
The problem is that the interpolation/extrapolation mechanism can change results quite significantly, so am pretty sure the results you get differ coz of this now (the major source of issues was not using the masked median though)

alistairsellar commented:
What do you get if you apply the median in masked mode but leave the interpolator as linear? For comparison with the original this would be the most direct comparison. 
We can then discuss as a separate issue whether Nearest is more correct.

valeriupredoi commented:
I get:
```
permafrost area,13.54527836534327
fraction area permafrost over zerodeg,0.29150328299094397
offset 1m minus surface,0.9465637
offset surface minus air,7.673065
attenuation 1m over surface,0.49492978649195407
attenuation surface over air,0.4920951365355539
```
with the setup identical to my previous run that we compared with

alistairsellar commented:
OK, thanks for trying. I reckon I need to dig in a bit with print statements then. Are you still up for making an executable script that I could run on Jasmin?

valeriupredoi commented:
sorry, got distracted with other shtuffs :giraffe: try this: on JASMIN: in a completely clean environment (no sys vars declared, no Python virtual envs activated):
```
eval "$(/home/users/valeriu/anaconda3/bin/conda shell.bash hook)"
conda activate release220
git clone https://github.com/ESMValGroup/ESMValTool.git
cd ESMValTool
git checkout autoassess_landsurface_recipes
pip install -e .[develop]
```
set up your config file in `~/.esmvaltool` to be like mine in `/home/users/valeriu/.esmvaltool/config-user.yml` (may as well copy it) and the you should be able to run, and while running edit the script in your local copy of `ESMValTool/esmvaltool/diag_scripts/autoassess/land_surface_permafrost/permafrost.py`

alistairsellar commented:
Thanks V. I get the following error from the pip command:
'''
    The following error occurred while trying to add or remove files in the
    installation directory:
    
        [Errno 13] Permission denied: '/home/users/valeriu/anaconda3/envs/release220/lib/python3.9/site-packages/test-easy-install-28528.write-test'
    
    The installation directory you specified (via --install-dir, --prefix, or
    the distutils default setting) was:
    
        /home/users/valeriu/anaconda3/envs/release220/lib/python3.9/site-packages/
 '''

alistairsellar commented:
Full output attached in case the above doesn't point to an obvious mistake.
[error.log](https://github.com/ESMValGroup/ESMValTool/files/6496110/error.log)

Presumably I made some schoolboy error?

valeriupredoi commented:
@alistairsellar - no, that's actually my schoolboy mistake for forgetting that pip needs to write to the root site-packages even if you perform a local, editable installation, can you run the pip install command but `--user` flag added to it? Sorry about that :relaxed:

valeriupredoi commented:
found another bug (self-bug, mind you) - the Koven sites locations need to be adjusted for lon: 0-360 data (they come on a -180-180 deg grid) so the interpolations can be correct! You may want to check that too in native Autoassess, @alistairsellar but I think all there is on a -180-180 deg grid. With the coordinates I am getting yet another set of metrics:
```
permafrost area,13.54527836534327
fraction area permafrost over zerodeg,0.2964807079171143
offset 1m minus surface,0.982940673828125
offset surface minus air,1.694000244140625
attenuation 1m over surface,0.6622823246301075
attenuation surface over air,0.5087502180308694
```
with masked medians, and
```
permafrost area,13.54527836534327
fraction area permafrost over zerodeg,0.2964807079171143
offset 1m minus surface,2.28289794921875
offset surface minus air,10.784530639648438
attenuation 1m over surface,0.6667226131008552
attenuation surface over air,0.5640653137761138
```
without masked medians! This thing is sensitive like heck on masks and interpolation mechanisms (as it should be :grin: )

alistairsellar commented:
> can you run the pip install command but with --user flag added to it?

Thanks V. I tried that now (`pip install -e .[develop] --user`), and I'm afraid I get the same error message. Should I try manually setting `--install-dir` to override what it's trying to use?

valeriupredoi commented:
ugh, try that (altho I've never used that option), alternatively, it's time you installed it fully in your JASMIN home :computer:

alistairsellar commented:
Sorry, choka with meetings all day. I'll try it now. 

And yes you're right it's high time I did my own proper installation. Off to the tutorial for me...

valeriupredoi commented:
no worries, I been to get jabbed :syringe: - yes, it'll be much more convenient if you have your own installation, you can port these metrics at a crazy speed :grin:

valeriupredoi commented:
frenly ping @alistairsellar about new [rseults](https://github.com/ESMValGroup/ESMValTool/pull/2170#issuecomment-844006636) :beer:

alistairsellar commented:
Thanks V, that's great. Sorry - another bad day for me. Tomorrow looks better...

valeriupredoi commented:
no worries, man, get some rest! :beer:

alistairsellar commented:
Sorry for the radio silence. I've now got a personal install working and have reproduced your results for the permafrost metrics - so I should be able to debug that now. 

But first I'll take a look at the metrics for the other sections that you made yesterday...

alistairsellar commented:
...OK, here's what I reckon:
`surfrad` is close enough.
```
SurfRadNLW MedAbsErr ann,3.8124046325683594
SurfRadNLW MedAbsErr djf,5.2260589599609375
SurfRadNLW MedAbsErr jja,5.695487976074219
SurfRadNLW MedAbsErr mam,4.456073760986328
SurfRadNLW MedAbsErr son,4.6931610107421875
SurfRadNSW MedAbsErr ann,4.9277801513671875
SurfRadNSW MedAbsErr djf,6.69140625
SurfRadNSW MedAbsErr jja,6.107902526855469
SurfRadNSW MedAbsErr mam,4.683586120605469
SurfRadNSW MedAbsErr son,5.502941131591797
```
(Sorry for the change in order)
Differences 5% or less, which could be explained by differences processing (e.g. regridding)

alistairsellar commented:
Snow is not quite right:
```
snow MedAbsErr djf,5.917015075683594
snow MedAbsErr mam,4.424624443054199
snow MedAbsErr son,0.43985575437545776
```
mam in particular is out by >50%. I'll dig into this after the permafrost.

alistairsellar commented:
`soilmoisture` is way out:
```
soilmoisture MedAbsErr djf,0.07084117829799652
soilmoisture MedAbsErr jja,0.0563998743891716
soilmoisture MedAbsErr mam,0.0671340823173523
soilmoisture MedAbsErr son,0.06610548496246338
```
but I know why. I got an answer back from our local expert on how to deal with the missing variables. I now know how to convert mrso into the units needed for comparison with the obs, and will code this up after the others.

alistairsellar commented:
In short, leave these with me and I'll let you know how I get on...

alistairsellar commented:
Permafrost is now OK, including the offset metrics.

AutoAssess (including the bugfix noted below):
```
attenuation 1m over surface,0.49573169134648043
attenuation surface over air,0.4925018141311794
fraction area permafrost over zerodeg,0.28940865677944033
offset 1m minus surface,0.94659424
offset surface minus air,7.7096863
permafrost area,13.561272172616677
```
ESMValTool (after the above two commits):
```
attenuation 1m over surface,0.49573173498262546
attenuation surface over air,0.4920951365355539
fraction area permafrost over zerodeg,0.29002045066872484
offset 1m minus surface,0.9465637
offset surface minus air,7.673065
permafrost area,13.54527836534327
```

The fixes needed were the above two commits, plus there was a bug in AutoAssess that you must have fixed in the ESMValTool code when you ported it: using np.median() instead of np.ma.median() on masked arrays in koven_temp_offsets().

alistairsellar commented:
I converted mrsos to volumetric soil moisture following some advice from the original author of the AutoAssess code. That makes the metrics agree...

AutoAssess:
```
soilmoisture MedAbsErr djf,0.07084117829799652
soilmoisture MedAbsErr mam,0.0671340823173523
soilmoisture MedAbsErr jja,0.0563998743891716
soilmoisture MedAbsErr son,0.06610548496246338
```

ESMValTool:
```
soilmoisture MedAbsErr djf,0.0708264410495758
soilmoisture MedAbsErr mam,0.06645134091377258
soilmoisture MedAbsErr jja,0.05707845464348793
soilmoisture MedAbsErr son,0.06563765555620193
```

Surfrad was already OK in your test above, and permafrost and soil moisture are now sorted, so just snow to go. I should be able to fix that today.

valeriupredoi commented:
wohooo - awesome man, these metrics now look absolutely spot-on and I'm happy you didn't have to refactor too much :grin: Excellent, all that's left to be done now is to create the documentation for these metrics and upload the (few) figures, am afraid that's mostly you since I haven't the foggiest what to write. Let me know what I can do to help :beer:

valeriupredoi commented:
we need to find the equivalent of `mrsol` for CMIP5 - this is why the test is failing, it can't find the CMIP5 CMOR table for `mrsol` - we may have to create a derived variable? @jvegasbsc @zklaus  do you know what's the map of that? - I couldn't find it in the usual CMIP5->CMIP6 mapping dict we use

zklaus commented:
Look at [the CMIP6 data request](http://clipc-services.ceda.ac.uk/dreq/u/591389b8-9e49-11e5-803c-0d0b866b59f3.html) and [the CMIP5 standard output](https://pcmdi.llnl.gov/mips/cmip5/docs/standard_output.pdf?id=21) it seems that `mrsol` is the CMIP6 replacement for the CMIP5 variable `mrlsl`, possibly to homogenize the nomenclature among the soil moisture variables.

valeriupredoi commented:
shweet, cheers for the superquick answer @zklaus :beer: - @alistairsellar what say you? Also, for documentation purposes, for the new recipes, I have included a template in this [commit](https://github.com/ESMValGroup/ESMValTool/pull/2170/commits/6e9b7506d96ed3ad4eb1d0699fd8e7ddf5b4d15b) whic is a stripped-down version of the one we did for Stratosphere, so it's easier to copy along and modify for each of the four recipes we need to include :+1:

alistairsellar commented:
Yes, I agree that mrlsl = mrsol. How do you implement this kind of flexibility?

alistairsellar commented:
Thanks for the documentation template V - leave that with me. I can pull the relevant bits out of AutoAssess.

I'm going to try and get the snow metrics right first though.

valeriupredoi commented:
> Yes, I agree that mrlsl = mrsol. How do you implement this kind of flexibility?

OK, we'll have to write a new recipe only for CMIP5 data I reckon

alistairsellar commented:
Update: I've got to the bottom of the snow difference, and it lies with the input data. Looking at a few sample points, `snw` in LImon has much smaller values (e.g. factor of 4) than the snow variable used by AutoAssess. I've contacted our local snow expert, but since it relates to something quite fundamental, I think that we should plough (pun intended) ahead without it. Are you OK for me to delete the snow recipes from this branch?

alistairsellar commented:
@valeriupredoi I've updated the battle plan in the PR description. Is anything missing? I'm not sure what the "missing plots" one was about - is it still a thing?

valeriupredoi commented:
> @valeriupredoi I've updated the battle plan in the PR description. Is anything missing? I'm not sure what the "missing plots" one was about - is it still a thing?

@alistairsellar great, cheers! OK, let's forget about `snow`, summer's here anyway :grin: The missing plots in gallery bit was to remind us to add the plots to recipe figures (I think only `permafrost` outputs plots anyway, but we could add the metrics plots, if we run the recipes with two different models, probably would be nice to have those too?). And yes, we should remove the MOHC test recipes too (I've updated the battle plan as well)

alistairsellar commented:
Hey @valeriupredoi, I think I've done everything that I know how to! If there are outstanding tasks, please tell me.

Key updates:
- Documentation for permafrost, soil moisture & surface radiation now complete, including metrics plots and permafrost maps
- Fixed the permafrost recipe that was failing due to mrsol missing, by using mrsos instead (as you suggested at one point I think). Results unaffected due to the limited way in which this field is used.

alistairsellar commented:
I've ticked off most items in the PR description. Outstanding are:
- closes issue: should I open an issue retrospectively, or reference #1931?
- code quality guidelines: is it enough that the automated tests pass?
- recipe is well documented: given that I'm the one adding the documentation, this should be reviewed by someone else
- provenance information: I think I need technical help on this one

valeriupredoi commented:
Fantastic work @alistairsellar :beer: :tophat: 

> I've ticked off most items in the PR description. Outstanding are:
> 
>     * closes issue: should I open an issue retrospectively, or reference #1931?
> 

referencing it is enough, we should keep that open for the next stuffs (inc. snow which is under snow now)

>     * code quality guidelines: is it enough that the automated tests pass?
> 

that's perfect, documentation builds fine and we don't really have functional tests for recipes anyway (yet)

>     * recipe is well documented: given that I'm the one adding the documentation, this should be reviewed by someone else
> 

good point, shall we say @zklaus maybe - since he's the release manager anyway?

>     * provenance information: I think I need technical help on this one

we should just see what sort of provenance info we provided for Stratosphere, that should suffuce I believe -> @bouweandela what do you think?

valeriupredoi commented:
I changed the PR from Draft to RfR, @zklaus would you be a peach and have a look over the documentation pls mate :peach: Ali, it's you the technical/scientific reviewer :grin:

alistairsellar commented:
Fab, thanks V!

> we should just see what sort of provenance info we provided for Stratosphere

Sounds sensible - please can you point me to that? Sorry for the basic question.

alistairsellar reviewed (approved):
Approving science/technical aspects.

Review comments:
- Technical correctness: the technical interface layer used here has been employed previously for AutoAssess stratosphere diagnostic and works correctly for both that and the new diagnostics, as evidenced by the consistency with AutoAssess results when applied to the same runs / periods.
- Scientific robustness and maturity: the outputs of these diagnostics are consistent with those from AutoAssess, the code for which has been previously scientifically reviewed and used in model development for a number of years. With the exception of soil moisture (see below), no science changes in the code were required in the porting to ESMValTool (thanks to V's technical interface), so the science of the new ESMValTool code is mature.
- The only science change was to convert soil moisture content into volumetric soil moisture concentration without two Met Office variables that are not in CMIP. The new conversion for this was reviewed by Heather Rumbold on 2021-05-21: 
  > Yes I agree, the conversion from the kg/m2 units to m3/m3 - divide by the density of water (1000 kg/m3) and then divide by the layer depths/thickness in metres.

My review doesn't cover documentation.

valeriupredoi commented:
@alistairsellar you're a star :star: For the references/provenance here's how the documentatuion tells you to [do it](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#adding-references) - I am not sure if we need to add any here, @bouweandela is the provenancespecialist (it's his baby after all) maybe he can help out here? :beer:

alistairsellar commented:
Provenance tags now added to the recipes. I think that completes everything from my side. 

Status summary:
- I've reviewed the science aspects, having run equivalent AutoAssess diagnostics to check for consistency of results (comparisons logged above).
- The documentation should ideally be reviewed by someone else, since I added this to ESMValTool (although it was based on the original AutoAssess documentation for these metrics). Documentation pages are:
   - [Permafrost](https://esmvaltool--2170.org.readthedocs.build/en/2170/recipes/recipe_autoassess_landsurface_permafrost.html)
   - [Surface radiation](https://esmvaltool--2170.org.readthedocs.build/en/2170/recipes/recipe_autoassess_landsurface_surfrad.html)
   - [Soil moisture](https://esmvaltool--2170.org.readthedocs.build/en/2170/recipes/recipe_autoassess_landsurface_soilmoisture.html)
- Technical review required

alistairsellar commented:
@remi-kazeroni noted that the documentation for these recipes was not following the template. They have now been reformatted to follow it much more closely. The only deviation from the template is an additional section at the end explaining about the AutoAssess wrapper script, which could be helpful to some users if they need to make changes.

remi-kazeroni reviewed (changes requested):
Thanks a lot for all your work @alistairsellar and @valeriupredoi! This is already in very good shape and the documentation follows the template well. 

Besides minor points, my main comment is that the [provenance ](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#recording-provenance-in-a-python-diagnostic-script) is currently not recorded and should be added to the diagnostic scripts.

Also, I couldn't reproduce exactly the figure `Permafrost_Metrics.png`. It seems to me that there is a scaling issue for the x-axis which extends from 0 to 200 in my case. The `permafrost area` and the `fraction area permafrost over zerodeg` seem to be correct assuming a factor 100 (% ?) while other quantities seem to be fractional ([0-1] ?). Could you please check that?

valeriupredoi commented:
as for provenance output - let me do that in a separate PR after we've merged this one, not all diags have provenance output set, so it's not a make or break, let's just get this one over the hill and then I'll add it for all the AA diags (incl Stratosphere which is currently missing it too)

alistairsellar commented on doc/sphinx/source/recipes/recipe_autoassess_landsurface_surfrad.rst:
```diff
@@ -0,0 +1,182 @@
+---------------------------------
+
+2000-2009 climatologies (seasonal means) from CERES-EBAF Ed2.7.
+
```
Yes - my mistake in the documentation. Now corrected.

alistairsellar commented on doc/sphinx/source/recipes/recipe_autoassess_landsurface_permafrost.rst:
```diff
@@ -0,0 +1,204 @@
+.. figure:: /recipes/figures/autoassess_landsurface/pf_extent_asia_UKESM1-0-LL.png
+   :scale: 50 %
+   :alt: pf_extent_asia_UKESM1-0-LL.png
+
```
Good point. I've now switched these sample plots to ACCESS-CM2.

alistairsellar commented on esmvaltool/recipes/recipe_autoassess_landsurface_permafrost.yml:
```diff
@@ -0,0 +1,66 @@
+  description: |
+    Recipe that runs the Autoassess Land-surface assessment area diagnostic.
+    Land surface area and ice masks are time independent.
+    
```
Yes it should. Now added.

alistairsellar commented:
Brilliant - thanks @remi-kazeroni!

I think that all of the points have been answered now, apart from this one:
> Also, I couldn't reproduce exactly the figure `Permafrost_Metrics.png`. It seems to me that there is a scaling issue for the x-axis which extends from 0 to 200 in my case. The `permafrost area` and the `fraction area permafrost over zerodeg` seem to be correct assuming a factor 100 (% ?) while other quantities seem to be fractional ([0-1] ?). Could you please check that?

That's odd. I've just re-run this recipe from the latest revision and the plot is the same as https://esmvaltool--2170.org.readthedocs.build/en/2170/_images/Permafrost_Metrics.png. Please could you send or point me to your metrics.csv files from the output dirs? These should be
`plots/aa_landsurf_permafrost/autoassess_landsurf_permafrost/*/land_surface_permafrost/*/metrics.csv`.

remi-kazeroni commented:
> That's odd. I've just re-run this recipe from the latest revision and the plot is the same as https://esmvaltool--2170.org.readthedocs.build/en/2170/_images/Permafrost_Metrics.png. Please could you send or point me to your metrics.csv files from the output dirs? These should be
> `plots/aa_landsurf_permafrost/autoassess_landsurf_permafrost/*/land_surface_permafrost/*/metrics.csv`.

You're welcome @alistairsellar! I will review your changes now.

I have just rerun the `recipe_autoassess_landsurface_permafrost.yml` with the latest changes to this branch and still get the same behaviour. Here are the metrics files you asked (renamed *.txt to send them through GitHub). I hope I didn't do anything wrong to run this recipe.
![Permafrost_Metrics](https://user-images.githubusercontent.com/70641264/122183228-51052600-ce8b-11eb-9629-f43144fd5132.png)

[E3SM-1-0_metrics.csv.txt](https://github.com/ESMValGroup/ESMValTool/files/6661339/E3SM-1-0_metrics.csv.txt)
[ACCESS-CM2_metrics.csv.txt](https://github.com/ESMValGroup/ESMValTool/files/6661340/ACCESS-CM2_metrics.csv.txt)

remi-kazeroni reviewed (commented):
Thanks for the changes @valeriupredoi @alistairsellar! These look fine to me.

I noted that you will take care of the provenance in a sperated PR as autoassess diagnostics are a bit particular. I will approve this PR once the remaining issue witht the `Permafrost_Metrics.png` is clarified.

bouweandela commented:
> we should just see what sort of provenance info we provided for Stratosphere, that should suffuce I believe -> @bouweandela what do you think?

> as for provenance output - let me do that in a separate PR after we've merged this one, not all diags have provenance output set, so it's not a make or break, let's just get this one over the hill and then I'll add it for all the AA diags (incl Stratosphere which is currently missing it too)

Provenance is required for new diagnostics, the fact that you got away once without adding it doesn't mean it's not needed. Could you please add provenance in this pull request for those diagnostics that are new in this pull request?

valeriupredoi commented:
OK OK, I'll a function now :hammer_and_wrench:

bouweandela commented:
Thanks!

alistairsellar commented:
Thanks V! :beers: 2u

Let me know if you need any input information for the provenance, or any 🍺 .

valeriupredoi commented:
on it @alistairsellar - should be fine, I'll just output stuff that's already in the code, I'll ping you if I gets stuck :stop_button:

valeriupredoi commented:
dunit @alistairsellar - could you pls have a quick look at the provenance parameters (eg statistics, domains etc) if they're correct? If all's good I reckon we done and done :grin:

alistairsellar commented on esmvaltool/diag_scripts/autoassess/land_surface_soilmoisture/soilmoisture.py:
```diff
@@ -5,37 +5,58 @@
+        'caption': caption,
+        'statistics': ['mean'],
+        'domains': ['global'],
+        'plot_type': 'map',
```
soilmoisture diagnostic only produces the metrics plot, so 'map' isn't right. Not sure how best to describe the metrics plot, perhaps 'point', or 'bar'? Is there a list of allowed values for 'plot_type'?

valeriupredoi replied:
we can probably put `autoassess metric`? AFAIK there is no restricted list of possible values here

alistairsellar commented on esmvaltool/diag_scripts/autoassess/land_surface_surfrad/surfrad.py:
```diff
@@ -7,19 +7,44 @@
+        'caption': caption,
+        'statistics': ['mean'],
+        'domains': ['global'],
+        'plot_type': 'map',
```
surfrad diagnostic only produces the metrics plot, so 'map' isn't right. Not sure how best to describe the metrics plot, perhaps 'point', or 'bar'? Is there a list of allowed values for 'plot_type'?

valeriupredoi replied:
sames here, `autoassess metric` is nicely self-explanatory

alistairsellar replied:
As above.

alistairsellar reviewed (commented):
Brilliant, thanks @valeriupredoi! Magic!

The parameters all look correct to me, apart from plot_type for soilmoisture and surfrad - see the question in this review.

alistairsellar commented:
@remi-kazeroni thanks for the files. Looking into this now...
One question - where are you running it? Or perhaps more importantly, which ESGF node is it pointing to?

alistairsellar commented:
@remi-kazeroni, OK so your metrics.csv file for E3SM-1-0 is identical to mine down to the last digit, but ACCESS-CM2 differs in those two metrics by a factor of 100 
```
Me: 
    permafrost area,11.943403914017559
    fraction area permafrost over zerodeg,0.2789722958859204
You:
    permafrost area,0.11943403918041137
    fraction area permafrost over zerodeg,0.002789722959799026
```
(and is identical in all others).

These normalised assessment metrics are a ratio between the control and the experiment, so if one of those is out by 100 then so is the ratio, and hence the metrics plot.

So why is ACCESS-CM2 wrong in your test, just for these two metrics? Your hunch has got to be right that there is a fraction / percent mixup here. These are the only metrics that make use of the `sftlf` land area mask, so my _guess_ is that your ESGF node has ACCESS-CM2 `sftlf` as a fraction (range 0-1) while the Jasmin/CEDA node has it (correctly) as a percentage. Is it possible that ACCESS initially delivered `sftlf` with the wrong units and subsequently corrected it, and that Jasmin has a newer version of that data than the ESGF node you are pointing to?

remi-kazeroni commented:
> So why is E3SM-1-0 wrong in your test, just for these two metrics? Your hunch has got to be right that there is a fraction / percent mixup here. These are the only metrics that make use of the `sftlf` land area mask, so my _guess_ is that your ESGF node has E3SM-1-0 `sftlf` as a fraction (range 0-1) while the Jasmin/CEDA node has it (correctly) as a percentage. Is it possible that E3SM initially delivered `sftlf` with the wrong units and subsequently corrected it, and that Jasmin has a newer version of that data than the ESGF node you are pointing to?
> 
> [Edit - I got the models the wrong way round initially. E3SM is the one that is too low by 100.]

Thanks for your investigations @alistairsellar! I'm running the recipes at DKRZ which has an older version of the data you are referring to: `/mnt/lustre02/work/ik1017/CMIP6/data/CMIP6/CMIP/E3SM-Project/E3SM-1-0/historical/r1i1p1f1/fx/sftlf/gr/v20190919/sftlf_fx_E3SM-1-0_historical_r1i1p1f1_gr.nc`. That file contains indeed only [0-1] values... But I see that the version `v20201015` is made available on Jasmin but not on Mistral. That must explain the discrepancies. Sorry for the inconvenience!

alistairsellar commented:
> That must explain the discrepancies.

Phew! That's good news.

alistairsellar commented:
So apart from the small matter that my last commit has broken the tests, I think we're done. Provenance now added, and metrics issue resolved. Anything I've missed?

I'll fix that final error now...

alistairsellar commented:
Test now fixed. Back to you @remi-kazeroni - I _think_ we've covered everything raised by your review.

valeriupredoi commented:
> [Edit (by Alistair): correcting the quote since I corrected my original post that it was quoting. Also, I think you meant lower on DKRZ, rather than higher. I don't know if it's bad form to edit someone else's post, but I didn't want to confuse Remi...]

good point and call, Alistair! I am working in 30C at home, I don't even know how me brain is not mush :hot_face:

remi-kazeroni reviewed (commented):
Thanks for adding the provenance @valeriupredoi & @alistairsellar! For each three recipes, a `diagnostic_provenance.yml` file is created. But 2 of these recipes crash before the end because of the provenance. There seems to be a last bug for soilmoisture and surfrad... Could you please have a look? I attached the `main_log_debug.txt` and `log.txt` for one of these. 
[log.txt](https://github.com/ESMValGroup/ESMValTool/files/6668338/log.txt)
[main_log_debug.txt](https://github.com/ESMValGroup/ESMValTool/files/6668339/main_log_debug.txt)

remi-kazeroni commented:
Forgot to ask: do you want to edit the PR's title or is it fine like this for the merge and the changelog?

valeriupredoi commented:
cheers @remi-kazeroni (sorry @alistairsellar I re-edited the title :grin: ) - I have also updated https://github.com/ESMValGroup/ESMValTool/issues/1931 that should not be closed yet since snow still needs attention but now it contains the correct pointers to the work here

valeriupredoi commented:
> Thanks for adding the provenance @valeriupredoi & @alistairsellar! For each three recipes, a `diagnostic_provenance.yml` file is created. But 2 of these recipes crash before the end because of the provenance. There seems to be a last bug for soilmoisture and surfrad... Could you please have a look? I attached the `main_log_debug.txt` and `log.txt` for one of these.
> [log.txt](https://github.com/ESMValGroup/ESMValTool/files/6668338/log.txt)
> [main_log_debug.txt](https://github.com/ESMValGroup/ESMValTool/files/6668339/main_log_debug.txt)

eek! I'll have a look now, cheers mate :beer:

valeriupredoi commented:
OK @remi-kazeroni - fixed it, silly code was trying to re-write the record when running the bit to plot the metrics, my bad, I should have tested by running the thing yesterday (I have now :grin: )

remi-kazeroni reviewed (approved):
Everything works fine, thanks a lot @alistairsellar and @valeriupredoi! I can run the three recipes successfully and the provenance is well written in each case. This is good to be merged in my opinion 🍺

valeriupredoi commented:
brilliant! cheers ever so much, Remi! since you helped so much with it, how about you merge it :medal_sports:

remi-kazeroni commented:
> 
> 
> brilliant! cheers ever so much, Remi! since you helped so much with it, how about you merge it 🏅

You're welcome @valeriupredoi! Thanks to you and @alistairsellar for all the work! Great to have this merged for the next release and the IS-ENES3 deliverable D9.2 👍

alistairsellar commented:
That's great! Thank you @valeriupredoi & @remi-kazeroni for your time on this! Great to be another step closer to using ESMValTool in operations at the UK Met Office.
